### PR TITLE
Magic syntax for constraints: `X > 5` -> `Gt(5)`, etc.

### LINKED ISSUE #28: Operator based `BaseMetadata` creation
Hi all,
I really like the idea of `BaseMetadata` to formulate `Annotated` types.

However, configuration using functions such as `Gt(10)` is not as easy as `x > 10` to read. It would be nice to provide a non-constraint class `All` and support methods such as `__gt__` like below.

```python
class All(BaseMetadata):
    def __gt__(self, other):
        return Gt(other)

X = All()  # could be better to keep `All` private and provide `X` publically.
Annotated[int, X > 5]  # equivalent to Annotated[int, Gt(5)]
```

Similar logic can by applied to construct several objects.

- `Interval` from `2 < X < 5`
- `MultipleOf` from `X % 3 == 0`
- `MaxLen` from `X.len() <= 10`

### PR DESCRIPTION
**Pull request by Zac-HD**
Closes #28; needs some more tests for validation errors and edge cases if we decide to go for it, and of course documentation.

Importantly, adding most of the magic to our existing `Interval` and `Len` classes means that we can keep the API surface minimal, without exposing 'incomplete constraint' objects that don't really mean anything.  Note that I've set this up to error out immediately if you try using the comparison magics on something which already represents a constraint.  You *can* still do e.g. `Interval(gt=3) < 5`; this seems weird but OK to me in that it's very clear what it does and there's no duplication of bounds.

### CODE REVIEW

**Comment by Zac-HD** [comment]
Pretty sad that `len()` refuses to return a non-integer... I guess we could have a `len_X` magic object too, which I like a little more than `X.len()`?

File: annotated_types/__init__.py [hunk 1]
```diff
@@ -205,6 +205,28 @@ def __iter__(self) -> Iterator[BaseMetadata]:
         if self.le is not None:
             yield Le(self.le)
 
+    def __check(self, other, which):
```
Comment (samuelcolvin) [inline annotated_types/__init__.py:208]:
```suggestion
    def __check(self, other: Literal["lo", "hi"], which: Literal["lo", "hi"]) -> None:
```

Is that right? Wither way, a type hint here would be useful for readers.

File: annotated_types/__init__.py [hunk 2]
```diff
@@ -205,6 +205,28 @@ def __iter__(self) -> Iterator[BaseMetadata]:
...
+        if (which == "lo" and (self.lt is not None or self.le is not None)) or (
+            which == "hi" and (self.gt is not None or self.ge is not None)
+        ):
+            raise ValueError(f"{self!r} is incompatible with `=={other!r}`")
+
+    def __lt__(self, other) -> "Interval":
```
Comment (samuelcolvin) [inline annotated_types/__init__.py:214]: again a type hint on `other` would be nice, same below.

**Review by samuelcolvin — COMMENTED** [review]
This looks great. Some more tests, e.g. for `__check` failing would be nice, I'm a bit unclear on the cases where the error is raised.

The final question is what we call `X`? I can see the argument for `X`, but I can also imagine it presence could be pretty confusing without usage to explain what it does.

One option would be to use ellipsis, so the usage becomes:

```py
Annotated[int, ... <= 4]
```

I think that's just as clear as `X` without requiring an extra input. Are there any other candidates from builtins that would work better?

**Comment by samuelcolvin** [comment]
Oh, :facepalm: `Annotated[int, ... <= 4]` can't have an implementation in this package, it would need to be another "we recommend you interpret..." like we previously had with `slice`.

Overall, I'm a bit torn on this.

This brings us back to @adriangb's early suggestion that we provide a method to collect annotations, or at least a method to call on a given annotation to interpret it, or we just give an example.

**Comment by samuelcolvin** [comment]
I think I prefer `len(X)`, or we could do `X.len()`, but it's less obvious.

**Comment by adriangb** [comment]
I like the idea of shorthands like this, but every time I’ve implemented them it’s come back to bite me because of edge cases, complexity or just going unused. I also had to read this a couple times before I understood what was going on. Personally I would pause until Pydantic V2 goes live and we get more feedback from users. We can always add it at that point without even having to change Pydantic!

**Comment by samuelcolvin** [comment]
Agreed.
